### PR TITLE
feat: remove dedicated runtime for grpc, mysql and pg protocols 

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

Refine runtime related usages.

Major changes:
- remove dedicated runtime for gRPC, MySQL and Postgres
- replace read, write, bg runtimes with the new unified one `global_runtime`
- add a new `compact_runtime` for compaction only to limit its max CPU usage. This is a simple workaround at present.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

### CODE REVIEW

File: src/common/runtime/src/global.rs [hunk 1]
```diff
@@ -95,48 +90,38 @@ macro_rules! define_spawn {
...
-                .unwrap_or_else(|| create_runtime("global-read", "read-worker", READ_WORKERS)),
-            write_runtime: write
-                .unwrap_or_else(|| create_runtime("global-write", "write-worker", WRITE_WORKERS)),
-            bg_runtime: background
-                .unwrap_or_else(|| create_runtime("global-bg", "bg-worker", BG_WORKERS)),
+            global_runtime: global
+                .unwrap_or_else(|| create_runtime("global", "global-worker", GLOBAL_WORKERS)),
+            compact_runtime: compact
+                .unwrap_or_else(|| create_runtime("compact", "compact-worker", COMPACT_WORKERS)),
             hb_runtime: heartbeat
```
Comment (zhongzc) [inline src/common/runtime/src/global.rs:103]: It turns out that `hb` means `heartbeat`. How about making it clear?

File: src/servers/tests/mysql/mysql_server_test.rs [hunk 2]
```diff
@@ -45,14 +46,13 @@ struct MysqlOpts<'a> {
...
-    let query_handler = create_testing_sql_query_handler(table);
-    let io_runtime = Arc::new(
-        RuntimeBuilder::default()
-            .worker_threads(4)
-            .thread_name("mysql-io-handlers")
-            .build()
-            .unwrap(),
-    );
+    let query_handler: Arc<dyn SqlQueryHandler<Error = servers::error::Error> + Send + Sync> =
+        create_testing_sql_query_handler(table);
```
Comment (zhongzc) [inline src/servers/tests/mysql/mysql_server_test.rs:50]:
```suggestion
let query_handler = create_testing_sql_query_handler(table);
```

**Review by zhongzc — APPROVED** [review]

File: config/datanode.example.toml [hunk 3]
```diff
@@ -73,11 +73,9 @@ watch = false
 ## The runtime options.
 [runtime]
 ## The number of threads to execute the runtime for global read operations.
```
Comment (killme2008) [inline config/datanode.example.toml:75]: Don't forget to update the configuration comments

File: src/common/runtime/src/global.rs [hunk 4]
```diff
@@ -95,48 +90,38 @@ macro_rules! define_spawn {
...
         Self {
-            read_runtime: read
-                .unwrap_or_else(|| create_runtime("global-read", "read-worker", READ_WORKERS)),
-            write_runtime: write
-                .unwrap_or_else(|| create_runtime("global-write", "write-worker", WRITE_WORKERS)),
-            bg_runtime: background
-                .unwrap_or_else(|| create_runtime("global-bg", "bg-worker", BG_WORKERS)),
+            global_runtime: global
+                .unwrap_or_else(|| create_runtime("global", "global-worker", GLOBAL_WORKERS)),
+            compact_runtime: compact
```
Comment (sunng87) [inline src/common/runtime/src/global.rs:101]: if we are adding something like async index, will we reuse this runtime?

**Comment by sunng87** [comment]
Actually this is a breaking change. It changes how we configure and tune the server. We'd better including this in next minor release instead of patch release.

**Comment by waynexia** [comment]
> Actually this is a breaking change. It changes how we configure and tune the server. We'd better including this in next minor release instead of patch release.

I've tested this change won't fail the server start progress. Only the old config won't take effect so I didn't categorize it as breaking change.

File: config/datanode.example.toml (hunk 3 above)
Comment (waynexia) [reply config/datanode.example.toml:75]: It's updated

File: src/common/runtime/src/global.rs (hunk 1 above)
Comment (waynexia) [reply src/common/runtime/src/global.rs:103]: I changed the runtime name. The thread name has char limits

File: src/common/runtime/src/global.rs (hunk 4 above)
Comment (waynexia) [reply src/common/runtime/src/global.rs:101]: Yes, this global one is for all operations. Two exceptions are compaction and heartbeat (not included at present

**Review by evenyag — APPROVED** [review]

File: src/datanode/src/alive_keeper.rs [hunk 5]
```diff
@@ -187,7 +187,7 @@ impl RegionAliveKeeper {
             let running = self.started.clone();
 
             // Watches changes
-            common_runtime::spawn_bg(async move {
+            common_runtime::spawn_global(async move {
```
Comment (evenyag) [inline src/datanode/src/alive_keeper.rs:190]:
```suggestion
            common_runtime::spawn_hb(async move {
```

File: src/datanode/src/alive_keeper.rs [hunk 6]
```diff
@@ -286,7 +286,7 @@ impl CountdownTaskHandle {
             region_id,
             rx,
         };
-        let handler = common_runtime::spawn_bg(async move {
+        let handler = common_runtime::spawn_global(async move {
```
Comment (evenyag) [inline src/datanode/src/alive_keeper.rs:289]:
```suggestion
        let handler = common_runtime::spawn_hb(async move {
```

**Review by evenyag — COMMENTED** [review]
Do we need to run the alive keeper in the heartbeat runtime? The region alive keeper may be related to the heartbeat task.

**Comment by evenyag** [comment]
Still timeout
https://github.com/GreptimeTeam/greptimedb/actions/runs/10146099115/job/28053577040?pr=4436

**Comment by WenyXu** [comment]
> Still timeout https://github.com/GreptimeTeam/greptimedb/actions/runs/10146099115/job/28053577040?pr=4436

I updated the runtime config for tests. By default, each pod will be allocated up to 1 core CPU (which will lead the default global_rt_size to be 1). However, it's problematic to set the global runtime size to 1 (e.g., deadlock).

**Comment by evenyag** [comment]
> I updated the runtime config for tests. By default, each pod will be allocated up to 1 core CPU (which will lead the default global_rt_size to be 1). However, it's problematic to set the global runtime size to 1 (e.g., deadlock).

I found the table is flushing:
```
2024-07-29T14:29:49.385378397Z stdout F 2024-07-29T14:29:49.385189Z  INFO operator::request: Handle table manual flush request: FlushTableRequest { catalog_name: "greptime", schema_name: "public", table_name: "quod" }
```

The admin function may be the cause. We try to block inside the async runtime and hangs the whole runtime.
https://github.com/GreptimeTeam/greptimedb/blob/53fc14a50b192f628f4ba6cfc759de9e1d063105/src/common/macro/src/admin_fn.rs#L230-L248

File: src/common/macro/src/admin_fn.rs [hunk 7]
```diff
@@ -228,7 +228,7 @@ fn build_struct(
                         .create_mutable_vector(rows_num);
 
                     if columns_num == 0 {
-                        let result = common_runtime::block_on_read(async move {
```
Comment (evenyag) [inline src/common/macro/src/admin_fn.rs:231]: The current implementation of the admin fn may hang the runtime.

**Review by tisonkun — APPROVED** [review]

**Pull request merged**